### ISSUE TITLE
Looking for a way to pass unencoded arbitrary values to android instrumentation runner

### ISSUE THREAD

**Issue by mikehardy** [wix/Detox#2933]
Hi there, thanks for Detox, I use it a bunch.

I have a need to pass a parameter directly to the instrumentation runner launched as part of an android detox test.

https://firebase.google.com/docs/app-check/android/debug-provider#ci (steps 4.b to set and 5 where it is used - normally the build.gradle instrumentation arg would work, but Detox starts the instrumentation directly via adb, so I need to use Detox launch args passing method)

It is not one of the set of typical reserved instrumentation arguments (which are a bit out of date vs the current instrumentation runner, by the way...)

The argument itself is arbitrary from the perspective of Detox, there's no reason Detox should have knowledge of this launch argument

**Describe the solution you'd like**

A simple solution might be to have a special prefix users may append to any launch arg which signals to Detox's launch argument processor that this should not be encoded, and the key minus prefix plus the un-encoded variable should be sent directly

e.g., use a launch arg like `detoxInstrumentationArg<my arg name here>` and while it is verbose it is also unambiguous what we intend with the arg, so we could remove the prefix and pass it in with an `if (key.startsWith('detoxInstrumentationArg'))`in the middle here that stripped the prefix and added to a new warning perhaps, indicating that an arg was sent direct to instrumentation vs to the app:

https://github.com/wix/Detox/blob/9f158995e0c00b06a1b589021dc064bb82d5958d/detox/src/devices/drivers/android/tools/instrumentationArgs.js#L14-L20

**Describe alternatives you've considered**

Perhaps splitting the idea of instrumentation args and app args completely on Android, that would model the system better, but I think it would also be less compatible with regard to ios parity, and it would potentially be a breaking change

In the meantime I'm either hacking my argument directly in to `ADB.js` `startInstrumentation` via patch-package (functional, but yucky) or I'm using Java reflection to do something ugly and bypass the argument-passing by directly setting a value into Java (functional, but yucky)

**Additional context**

I could propose a PR if the idea is acceptable

**Comment by d4vidi** [comment]
Hi again @mikehardy! 😄
Truly happy to hear you like using Detox.
It's as you said it - Detox doesn't support user-specified instrumentation args, ATM. Sorry you had to patch things to get that done.

My instinct is to just say that we could accept a PR that would make `instrumentationLaunchArgs` a valid (pojo) parameter to handle in `device.launchApp()`. Unfortunately, funny as it may be, we have some changes planned out - with work already under way, to support multiple-apps (see #2840) in a way that wouldn't allow that. The arguments would therefore could only be supported coming from a more static, preliminary source, such as definition of apps / configuration in the Detox config file (in fact, `launchArgs` [can be defined](https://github.com/wix/Detox/blob/master/docs/APIRef.LaunchArgs.md#arguments-setup) there today, already). Would that suit your needs? If so, it sounds like that is PR we _could_ definitely accept.

>It is not one of the set of typical reserved instrumentation arguments (which are a bit out of date vs the current instrumentation runner, by the way...)

Current implementation seems aligned with Google's [documentation[(https://developer.android.com/studio/test/command-line#AMOptionsSyntax). Am I missing something here?

**Comment by mikehardy** [comment]
So, to be clear, if the 'instrumentationLaunchArgs' were set in the config it would be okay? Or maybe from the environment (it's a shared secret going through) but similarly early? That would be fine by me

I believe it was related to coverage arguments and maybe test specification, that was the last time I was in there, I got native Java coverage working from detox runs. That's a tangent though of course, if I propose anything there I'd link to the docs for support

**Comment by d4vidi** [comment]
Yes - setting it in the configuration, side-by-side with `launchArgs`, would be ok. I'd rather not to allow it from the environment, seeing that Detox config is complex as it is - so I'd rather go with the existing paradigms.

**Comment by d4vidi** [comment]
Happy to help. Send us your email (e.g. DM on [twitter](https://twitter.com/detoxe2e)) and I'll send an invite to our open Slack, so we could collaborate more easily if needed 👍🏻

**Comment by d4vidi** [comment]
Thanks
